blazar/tests/enforcement/test_enforcement.py

```diff
@@ -0,0 +1,209 @@
+            'service_name': 'compute1',
+            'vcpus': 4,
+            'cpu_info': 'foo',
+            'hypervisor_type': 'xen',
+            'hypervisor_version': 1,
+            'memory_mb': 8192,
+            'local_gb': 10,
+        }
+
+    def tearDown(self):
+        super(EnforcementTestCase, self).tearDown()
+
+    def test_load_filters(self):
+        print(self.enforcement.enabled_filters)
+        self.assertEqual(len(self.enforcement.enabled_filters), 1)
+
+        fake_filter = self.enforcement.enabled_filters[0]
+
+        self.assertEqual(fake_filter.conf.enforcement.fake_opt, 1)
+
+    def test_format_context(self):
+
+        formatted_context = self.enforcement.format_context(context.current())
+
+        expected = dict(user_id=111, project_id=222, region_name=self.region,
+                        auth_url='https://fakeauth.com')
+
+        self.assertDictEqual(expected, formatted_context)
+
+    def test_check_create(self):
+        allocation_candidates = {'virtual:instance': [get_fake_host('1')]}
+        lease_values = get_fake_lease()
+        reservations = list(lease_values['reservations'])
+
+        del lease_values['reservations']
+        ctx = context.current()
+
+        check_create = self.patch(self.enforcement.enabled_filters[0],
+                                  'check_create')
+
+        self.enforcement.check_create(ctx, lease_values, reservations,
+                                      allocation_candidates)
+
+        formatted_lease = self.enforcement.format_lease(
+            lease_values, reservations, allocation_candidates)
+        formatted_context = self.enforcement.format_context(ctx)
+
+        check_create.assert_called_once_with(formatted_context,
+                                             formatted_lease)
+
+    def test_check_update(self):
+        current_allocations = {'virtual:instance': [get_fake_host('1')]}
+        lease = get_fake_lease()
+        reservations = list(lease['reservations'])
+
+        new_lease_values = get_fake_lease(end_date=u'2014-02-07 13:37')
+        new_reservations = list(new_lease_values['reservations'])
+        allocation_candidates = {'virtual:instance': [get_fake_host('2')]}
+
+        del new_lease_values['reservations']
+        ctx = context.current()
+
+        check_update = self.patch(self.enforcement.enabled_filters[0],
+                                  'check_update')
+
+        self.enforcement.check_update(
+            ctx, lease, new_lease_values, current_allocations,
+            allocation_candidates, reservations, new_reservations)
+
+        formatted_context = self.enforcement.format_context(ctx)
+        formatted_lease = self.enforcement.format_lease(
+            lease, reservations, current_allocations)
+        new_formatted_lease = self.enforcement.format_lease(
+            new_lease_values, new_reservations, allocation_candidates)
+
+        check_update.assert_called_once_with(
```

Comment:
Similar concept to the above, I think what is really important here is that the lease is transitioned to the error state when the check_update fails, and succeeds otherwise. Mocking the check_update probably makes sense for that test.

Reply:
So I think we can't really test the transition in the service or enforcement unit tests. We have to rely on test_status unit tests that an exception raised will transition a lease to an error state. We can then test that exceptions do in fact get raised by enforcement checks and filter checks. Not sure if that makes a lot of sense.

```python
    def test_lease_status_func_raise_exception(self):
        lease = {
            'status': status.LeaseStatus.PENDING
        }
        lease_get = self.patch(self.db_api, 'lease_get')
        lease_get.return_value = lease
        lease_update = self.patch(self.db_api, 'lease_update')
        self.patch(self.status.LeaseStatus, 'is_valid_transition'
                   ).return_value = True

        @self.status.LeaseStatus.lease_status(
            transition=status.LeaseStatus.STARTING,
            result_in=(status.LeaseStatus.ACTIVE,))
        def dummy_start_lease(*args, **kwargs):
            raise exceptions.BlazarException

        self.assertRaises(exceptions.BlazarException,
                          dummy_start_lease,
                          lease_id=self.lease_id)

        lease_get.assert_called_once_with(self.lease_id)
        lease_update.assert_has_calls(
            [call(self.lease_id, {'status': status.LeaseStatus.STARTING}),
             call(self.lease_id, {'status': status.LeaseStatus.ERROR})])
```

Reply:
I see what you're saying... the status tests should account for testing what we want here.